src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
 
-            if (i == buffer.Length)
+            if (!EncodeHeaderName(name, buffer.Slice(i), out int nameLength))
             {
                 return false;
             }
+            i += nameLength;
 
-            if (!EncodeString(name, buffer.Slice(i), out int nameLength, lowercase: true))
+            if (!EncodeHeaderValue(value, values, separator, buffer.Slice(i), out int valueLength))
             {
                 return false;
             }
+            i += valueLength;
 
-            i += nameLength;
+            length = i;
+            return true;
+        }
 
-            if (i >= buffer.Length)
+        // Encode header name. It needs to be lower cased and validity was checked when created.
+        private static bool EncodeHeaderName(string s, Span<byte> buffer, out int nameLength)
+        {
+            const int ToLowerMask = 0x20;
+            nameLength = 0;
+
+            if (!IntegerEncoder.Encode(s.Length, 7, buffer, out int currentIndex))
             {
                 return false;
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            if (currentIndex + s.Length >= buffer.Length)
             {
                 return false;
             }
 
-            i += valueLength;
+            // TODO: use huffman encoding
+            for (int j = 0; j < s.Length; j++)
+            {
+                // TODO Use ASCII ToUpper when #34144 is ready.
+                buffer[currentIndex++] = (byte)(s[j] | (s[j] >= 'A' && s[j] <= 'Z' ? ToLowerMask : 0));
+            }
+            nameLength = currentIndex;
 
-            length = i;
             return true;
         }
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        // Encode fragment or header value without writing out length.
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int currentIndex)
         {
-            const int toLowerMask = 0x20;
+            if (s.Length >= buffer.Length)
+            {
+                return false;
+            }
 
+            int i = 0;
+            foreach (char c in s)
+            {
+                if ((c & 0xFF80) != 0)
+                {
+                    throw new HttpRequestException(SR.net_http_request_invalid_char_encoding);
+                }
+                buffer[i++] = (byte)c;
+            }
+            currentIndex += i;
+
+            return true;
+        }
+
+        private static bool EncodeHeaderValue(string s, string[] parts, string separator, Span<byte> buffer, out int length)
+        {
             int i = 0;
             length = 0;
 
-            if (buffer.Length == 0)
+            // \0, length and at least one byte of value
+            if (buffer.Length < 3)
             {
                 return false;
             }
 
-            buffer[0] = 0;
+            if (s != null)
+            {
+                i = s.Length;
```

Comment:
Instead of `i` can we call this something more descriptive, like `headerValueLength` or similar?